src/ceph.rs

```diff
@@ -104,53 +295,1287 @@ pub fn connect_to_ceph(user_id: &str, config_file: &str) -> Result<rados_t, Rado
+    if ctx.is_null() {
+        return Err(RadosError::new("Rados ioctx not created.  Please initialize first".to_string()));
+    }
+    let obj_name_str = try!(CString::new(object_name));
+
+    unsafe {
+        let ret_code = rados_write_full(ctx, obj_name_str.as_ptr(), buffer.as_ptr() as *const i8, buffer.len());
+        if ret_code < 0 {
+            return Err(RadosError::new(try!(get_error(ret_code as i32))));
+        }
+    }
+    Ok(())
+}
+
+/// Efficiently copy a portion of one object to another
+/// If the underlying filesystem on the OSD supports it, this will be a copy-on-write clone.
+/// The src and dest objects must be in the same pg. To ensure this, the io context should
+/// have a locator key set (see rados_ioctx_locator_set_key()).
+pub fn rados_object_clone_range(ctx: rados_ioctx_t, dst_object_name: &str, dst_offset: u64, src_object_name: &str,
+                                src_offset: u64, length: usize)
+                                -> Result<(), RadosError> {
+    if ctx.is_null() {
+        return Err(RadosError::new("Rados ioctx not created.  Please initialize first".to_string()));
+    }
+    let dst_name_str = try!(CString::new(dst_object_name));
+    let src_name_str = try!(CString::new(src_object_name));
+
+    unsafe {
+        let ret_code =
+            rados_clone_range(ctx, dst_name_str.as_ptr(), dst_offset, src_name_str.as_ptr(), src_offset, length);
+        if ret_code < 0 {
+            return Err(RadosError::new(try!(get_error(ret_code as i32))));
+        }
+    }
+    Ok(())
+}
+
+/// Append len bytes from buf into the oid object.
+pub fn rados_object_append(ctx: rados_ioctx_t, object_name: &str, buffer: &[u8]) -> Result<(), RadosError> {
+    if ctx.is_null() {
+        return Err(RadosError::new("Rados ioctx not created.  Please initialize first".to_string()));
+    }
+    let obj_name_str = try!(CString::new(object_name));
+
+    unsafe {
+        let ret_code = rados_append(ctx, obj_name_str.as_ptr(), buffer.as_ptr() as *const i8, buffer.len());
+        if ret_code < 0 {
+            return Err(RadosError::new(try!(get_error(ret_code as i32))));
+        }
+    }
+    Ok(())
+}
+
+/// Read data from an object.  This fills the slice given and returns the amount of bytes read
+/// The io context determines the snapshot to read from, if any was set by
+/// rados_ioctx_snap_set_read().
+pub fn rados_object_read(ctx: rados_ioctx_t, object_name: &str, fill_buffer: &mut [u8], read_offset: u64)
+                         -> Result<i32, RadosError> {
+    if ctx.is_null() {
+        return Err(RadosError::new("Rados ioctx not created.  Please initialize first".to_string()));
+    }
+    let object_name_str = try!(CString::new(object_name));
+
+    unsafe {
+        let ret_code = rados_read(ctx,
+                                  object_name_str.as_ptr(),
+                                  fill_buffer.as_mut_ptr() as *mut i8,
```

Comment:
Testing has revealed a problem with this.  Ceph returns -34 "unknown error".  Some gdb poking might reveal what is wrong.